### PR TITLE
Review of #690 by @olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
- This is a review of #690
- Creating a custom PR to make it easier to track the changes for future reference

## Summary by CodeRabbit

- **New Features**
	- Introduced a new relay difficulty management system based on target hashes instead of minimum difficulty bits, enhancing flexibility and precision in relay mining.
	- Added detailed telemetry and logging improvements for relay mining difficulty metrics.
	- Implemented a scaling function for relay difficulty target hashes, allowing for dynamic adjustments based on performance.
  
- **Bug Fixes**
	- Enhanced validation logic for relay difficulty parameters to ensure correctness in processing and updating values.

- **Tests**
	- Comprehensive updates to existing tests to align with the new relay difficulty target hash system, ensuring accuracy in functionality and behavior verification.

### CODE REVIEW

**Comment by Olshansk** [comment]
@bryanchriswhite @red-0ne Just FYI. Going to merge this in so we can get #690 over the finish line.

File: testutil/network/network.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func New(t *testing.T, configs ...Config) *Network {
 		cfg = configs[0]
 	}
 	net, err := network.New(t, t.TempDir(), cfg)
-	require.NoError(t, err)
+	require.NoError(t, err, "TODO_FLAKY: This config setup is periodically flakyis a flaky is ")
```
Comment (bryanchriswhite) [inline testutil/network/network.go:53]:
```suggestion
	require.NoError(t, err, "TODO_FLAKY: This config setup is periodically flaky")
```

File: proto/poktroll/proof/params.proto [hunk 2]
```diff
@@ -12,6 +12,7 @@ message Params {
   option (amino.name) = "poktroll/x/proof/Params";
   option (gogoproto.equal) = true;
 
+  // TODO_FOLLOWUP(@olshansk, #690): Either delete this or change it to be named "minimum"
```
Comment (red-0ne) [inline proto/poktroll/proof/params.proto:15]: Could you explain why this can be deleted?

Comment (red-0ne) [reply proto/poktroll/proof/params.proto:15]: I see now. Could this be a map of serviceId to difficulty target hash (string->bytes)?

File: x/tokenomics/keeper/settle_pending_claims.go [hunk 3]
```diff
@@ -104,6 +104,7 @@ func (k Keeper) SettlePendingClaims(ctx sdk.Context) (
 				expiredResult.NumComputeUnits += numClaimComputeUnits
 				continue
 			}
+			// TODO_FOLLOWUP(@olshansk, #690): Document the potential changes needed here.
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_pending_claims.go:107]: :eyes:

File: x/tokenomics/module/abci.go [hunk 4]
```diff
@@ -80,14 +81,16 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) (err error) {
...
-		copy(newTargetHash[:], newDifficulty.TargetHash)
+	// Telemetry - emit telemetry for each service's relay mining difficulty.
+	for serviceId, newRelayMiningDifficulty := range difficultyPerServiceMap {
+		var newRelayMiningTargetHash [protocol.RelayHasherSize]byte
+		copy(newRelayMiningTargetHash[:], newRelayMiningDifficulty.TargetHash)
 
-		difficulty := protocol.GetDifficultyFromHash(newTargetHash)
+		// NB: The difficulty integer is just a human readable interpretation of
+		// the target hash and is not actually used for business logic.
+		difficulty := protocol.GetDifficultyFromHash(newRelayMiningTargetHash)
```
Comment (red-0ne) [inline x/tokenomics/module/abci.go:91]: Don't we risk to have a difficulty that is greater than `uint64`?

File: pkg/crypto/protocol/hasher.go [hunk 5]
```diff
@@ -3,14 +3,13 @@ package protocol
...
+	RelayHasherSize      = sha256.Size
+	TrieHasherSize       = sha256.Size
+	TrieRootSize         = TrieHasherSize + trieRootMetadataSize
+	trieRootMetadataSize = 16 // TODO_CONSIDERATION: Export this from the SMT package.
 )
 
 var (
 	NewRelayHasher = sha256.New
-	NewTrieHasher = sha256.New
+	NewTrieHasher  = sha256.New
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply pkg/crypto/protocol/hasher.go:14]: This had gofmt issues and needs to be aligned. Rejecting suggestion.

File: proto/poktroll/proof/params.proto (hunk 2 above)
Comment (Olshansk) [reply proto/poktroll/proof/params.proto:15]: That's what the `RelayMiningDifficulty` on-chain type already is. See `/Users/olshansky/workspace/pocket/poktroll3/x/tokenomics/keeper/update_relay_mining_difficulty.go`.

```go
type RelayMiningDifficulty struct {
	ServiceId string `protobuf:"bytes,1,opt,name=service_id,json=serviceId,proto3" json:"service_id,omitempty"`
	BlockHeight int64 `protobuf:"varint,2,opt,name=block_height,json=blockHeight,proto3" json:"block_height,omitempty"`
	NumRelaysEma uint64 `protobuf:"varint,3,opt,name=num_relays_ema,json=numRelaysEma,proto3" json:"num_relays_ema,omitempty"`
	TargetHash []byte `protobuf:"bytes,4,opt,name=target_hash,json=targetHash,proto3" json:"target_hash,omitempty"`
}
```

The baseline is just `fffff`. 

Having an on-chain parameter for just extra overhead and context.

File: x/tokenomics/keeper/settle_pending_claims.go (hunk 3 above)
Comment (Olshansk) [reply x/tokenomics/keeper/settle_pending_claims.go:107]: #703

**Review by red-0ne — APPROVED** [review]
Let's merge it in! :rocket:

**Pull request merged**